compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/BinaryArithmeticNode.java

```diff
@@ -235,6 +235,165 @@ private static ReassociateMatch findReassociate(BinaryNode binary, NodePredicate
         return null;
     }
 
+    private static ReassociateMatch findReassociativeOperand(BinaryArithmeticNode<?> binary, NodePredicate criterion) {
+        NodePredicate binaryCriterion = new ReassociativePredicate(binary);
+        ValueNode forX = binary.getX();
+        ValueNode forY = binary.getY();
+        if (binaryCriterion.apply(forX) && findReassociate((BinaryNode) forX, criterion) != null) {
+            return ReassociateMatch.x;
+        }
+
+        if (binaryCriterion.apply(forY) && findReassociate((BinaryNode) forY, criterion) != null) {
+            return ReassociateMatch.y;
+        }
+        return null;
+    }
+
+    private static class ReassociativePredicate implements NodePredicate {
+        private final BinaryArithmeticNode<?> binary;
+
+        ReassociativePredicate(BinaryArithmeticNode<?> binary) {
+            this.binary = binary;
+        }
+
+        @Override
+        public boolean apply(Node node) {
+            if (binary instanceof AddNode || binary instanceof SubNode) {
+                return node instanceof AddNode || node instanceof SubNode;
+            }
+            return node.getClass() == binary.getClass();
+        }
+    }
+
+    private static class LoopPhiPredicate implements NodePredicate {
+        private final BinaryArithmeticNode<?> binary;
+
+        LoopPhiPredicate(BinaryArithmeticNode<?> binary) {
+            this.binary = binary;
+        }
+
+        @Override
+        public boolean apply(Node node) {
+            if (node instanceof ValuePhiNode) {
+                ValuePhiNode iv = (ValuePhiNode) node;
+                return iv.values().contains(binary);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Tries to expose values which satisfy the criterion as an operand. This could create more
+     * opportunities for other optimizations like constant-fold and loop-invariant elimination. For
+     * example with a constantness criterion: {@code (a * 2) * b => (a * b) * 2}
+     *
+     * This method accepts only {@linkplain BinaryOp#isAssociative() associative} operations such as
+     * +, -, *, &, | and ^
+     */
+    public static ValueNode reassociateUnmatchedValues(BinaryArithmeticNode<?> node, NodePredicate criterion, NodeView view, boolean skipLoopPhiValues) {
+        ValueNode forX = node.getX();
+        ValueNode forY = node.getY();
+        assert node.getOp(forX, forY).isAssociative();
+
+        // Skip re-associations for expressions like "i = i + var". This may influence the counted
+        // loop detection. For example, if the loop counter is re-associated from "i += (var + 1)"
+        // to "i = (i + var) + 1", the counted loop will not be detected as expected.
+        if (skipLoopPhiValues) {
+            LoopPhiPredicate phiCriterion = new LoopPhiPredicate(node);
```

Comment:
is the phi predicate class & allocation necessary here? I might miss something but it seems it does not escape and the logic of apply is very simple.

Reply:
Good catch! We could use a static function to check instead of `NodePredicate` here. Thanks!